R/class-workbook.R

```diff
@@ -4124,11 +4140,92 @@ wbWorkbook <- R6::R6Class(
       value
     },
 
+    validate_new_sheet = function(sheet) {
+      # returns nothing, throws error if there's a problem.
+
+      if (is.na(sheet)) {
+        stop("sheet cannot be NA")
+      }
+
+      if (is.numeric(sheet)) {
+        if (sheet %% 1 != 0) {
+          stop("If sheet is numeric it must be an integer")
+        }
+
+        if (sheet <= 0) {
+          stop("if sheet is an integer it must be a positive number")
+        }
+
+        if (sheet <= length(self$sheet_names)) {
+          stop("there is already a sheet at index ", sheet)
+        }
+
+        return()
+      }
+
+      sheet <- as.character(sheet)
+      if (any_illegal_chars(sheet)) {
+        stop("Illegal characters found in sheet. Please remove. See ?openxlsx::clean_worksheet_name")
+      }
+
+      if (nchar(sheet) > 31) {
+        stop("sheet names must be <= 31 chars")
+      }
+
+      if (tolower(sheet) %in% self$sheet_names) {
+        stop("a sheet with name '", sheet, '"already exists"')
+      }
+    },
+
+    get_sheet_index = function(sheet) {
+      # returns the sheet index, or NA
+      if (is.null(self$sheet_names)) {
+        warning("Workbook has no sheets")
+        return(NA_integer_)
+      }
+
+      if (is.character(sheet)) {
+        m <- match(tolower(sheet), tolower(self$sheet_names))
+        bad <- is.na(m)
+
+        if (any(bad)) {
+          stop("Sheet names not found: ", toString(sheet[bad]))
+        }
+
+        sheet <- m
+      } else {
+        sheet <- as.integer(sheet)
+        bad <- which(sheet > length(self$sheet_names))
+
+        if (length(bad)) {
```

Comment:
`length(bad) || sheet < 1` Not sure if we need to guard against nonsense, but maybe someone calculates sheet positions wrong ...?